qiskit_experiments/characterization/spectroscopy.py

```diff
@@ -0,0 +1,408 @@
+            - a standard deviation that is wider than the smallest frequency increment,
+            - a signal-to-noise ratio, defined as the amplitude of the peak divided by the
+              square root of the median y-value less the fit offset, greater than a
+              threshold of two, and
+            - a standard error on the sigma of the Gaussian that is smaller than the sigma.
+
+        Args:
+            fit_amp: Amplitude of the fitted peak.
+            fit_sigma: Standard deviation of the fitted Gaussian.
+            fit_freq: Frequency of the fitted peak.
+            fit_offset: Offset of the fit.
+            reduced_chisq: Reduced chi-squared of the fit.
+            xdata: x-values, i.e. the frequencies.
+            ydata: y-values, i.e. the measured signal.
+            fit_sigma_err: Errors on the standard deviation of the fit.
+
+        Returns:
+            computer_bad or computer_good if the fit passes or fails the criteria, respectively.
+        """
+        min_freq = xdata[0]
+        max_freq = xdata[-1]
+        freq_increment = xdata[1] - xdata[0]
+
+        snr = abs(fit_amp) / np.sqrt(abs(np.median(ydata) - fit_offset))
+        fit_width_ratio = fit_sigma / (max_freq - min_freq)
+
+        # pylint: disable=too-many-boolean-expressions
+        if (
+            min_freq <= fit_freq <= max_freq
+            and 1.5 * freq_increment < fit_sigma
+            and fit_width_ratio < 0.25
+            and reduced_chisq < 3
+            and (fit_sigma_err is None or (fit_sigma_err < fit_sigma))
+            and snr > 2
+        ):
+            return "computer_good"
+        else:
+            return "computer_bad"
+
+    @classmethod
+    def _format_plot(cls, ax, analysis_result):
+        """Format curve fit plot."""
+        ax.tick_params(labelsize=14)
+        ax.set_xlabel(f"Frequency ({analysis_result['unit']})", fontsize=16)
+        ax.set_ylabel("Signal [arb. unit.]", fontsize=16)
+        ax.grid(True)
+
+
+class Spectroscopy(BaseExperiment):
+    """Class that runs spectroscopy by sweeping the qubit frequency.
+
+    The circuits produced by spectroscopy, i.e.
+
+    .. parsed-literal::
+
+                   ┌────────────┐ ░ ┌─┐
+              q_0: ┤ Spec(freq) ├─░─┤M├
+                   └────────────┘ ░ └╥┘
+        measure: 1/══════════════════╩═
+                                     0
+
+    have a spectroscopy pulse-schedule embedded in a spectroscopy gate. The
+    pulse-schedule consists of a set frequency instruction followed by a GaussianSquare
+    pulse. A list of circuits is generated, each with a different frequency "freq".
+    """
+
+    __analysis_class__ = SpectroscopyAnalysis
+
+    # Supported units for spectroscopy.
+    __units__ = {"Hz": 1.0, "kHz": 1.0e3, "MHz": 1.0e6, "GHz": 1.0e9}
+
+    # default run options
+    __run_defaults__ = {"meas_level": MeasLevel.KERNELED, "meas_return": "single"}
```

Comment:
This should request `avg` by default otherwise you'll run into performance issue in practical situation. Usually this experiment is run as composite (parallel) experiment and requesting full iq points doesn't scale, because of the deserialization overhead (str-> float conversion in parser), and data volume itself (usually ~100 MB as qubit number scales). 
On the other hand, I agree it is good to generate the raw iq scatter plot along with fit result (when we get poor fitting usually there are lots of information we can get from iq scatter plot). This plot should be generated regardless of experiment type but I don't think we can efficiently perform, for example, QV experiment with level1-single setting in our cloud system. 
Is it possible to return raw iq plot image on backend side and attache this image to the experiment result? This enables us to check the detailed behavior of system without communicating huge text data of iq values (or add execution option `return_iq_plot`). @jyu00 @wshanks

Reply:
I don't think that the backends return the variance when averaging the single-shots. So we currently need single-shot data to get the errors. Note that if users will do this with a large number of qubits then these defaults can be overridden.